File: Array_And_Strings/2_checkPermutation.cpp
// Given 2 strings , check if one is permutaion of the other

#include<iostream>
using namespace std;

bool check(char* a, char* b, int m, int n)
{
	if(m!=n)
		return false;
	int alpha_count[256];
	for(int i=0;i<256;i++)
		alpha_count[i] = 0;
	for(int i=0;i<m;i++)
		alpha_count[(int)a[i]]++;
	for(int i=0;i<m;i++)
		alpha_count[(int)a[i]]--;
	for(int i=0;i<256;i++)
		if(alpha_count[i]!=0)
			return false;
	return true;
}

int main()
{
	int m,n;
	cout<<"Enter the length of string 1 : ";
	cin>>m;
	char a[m];
	cout<<"Enter string 1 : ";
	cin>>a;
	a[m] = '\0';
	cout<<"Enter the length of string 2 : ";
	cin>>n;
	char b[n];
	cout<<"Enter string 2 : ";
	cin>>b;
	b[n] = '\0';

	if(check(a,b,m,n))
		cout<<"Yes\n";
	else 
		cout<<"No\n";

	return 0;
}

File: Array_And_Strings/2_checkPermutation.java
// given 2 strings check if one is permutation of the other
// assuming it is an ASCII string
import java.util.Scanner;

class CheckPermutation
{
	
	boolean isPermutation(String a, int la, String b, int lb)
	{
		if(la!=lb)
			return false;

		char[] counts = new char[256];
		for(int i=0;i<la;i++)
		{
			counts[a.charAt(i)]++;
			counts[b.charAt(i)]--;
		}

		for(int i=0;i<la;i++)
			if(counts[i]!=0)
				return false;
		
		return true;
	}

	public static void main(String[] args)
	{
		Scanner in = new Scanner(System.in);
		CheckPermutation ob = new CheckPermutation();
		System.out.print("Enter the length of 1st string : ");
		int la = in.nextInt();
		System.out.print("Enter the 1st string : ");
		String a = in.next();
		System.out.print("Enter the length of 2nd string : ");
		int lb = in.nextInt();
		System.out.print("Enter the 2nd string : ");
		String b = in.next();

		System.out.println();
		if(ob.isPermutation(a,la,b,lb))
			System.out.println("Yes");
		else
			System.out.println("No");
	}

}
